### PR TITLE
filesystem: add rename component tests

### LINKED ISSUE #2014: filesystem: add component tests to test rename

### PR DESCRIPTION
**Pull request by talSofer**
closes #2014

### CODE REVIEW

File: clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java [hunk 1]
```diff
@@ -54,15 +54,14 @@
...
 import io.lakefs.clients.api.model.ObjectStats.PathTypeEnum;
-import io.lakefs.clients.api.model.ObjectStatsList;
-import io.lakefs.clients.api.model.Repository;
-import io.lakefs.clients.api.model.StagingLocation;
-import io.lakefs.clients.api.model.StagingMetadata;
 
 public class LakeFSFileSystemTest {
     protected static final Logger LOG = LoggerFactory.getLogger(LakeFSFileSystemTest.class);
+    private static final Long FILE_SIZE = 1L;
+    private static final Long MTIME = 0L;
```
Comment (arielshaqed) [inline clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java:62]:
```suggestion
    private static final Long UNUSED_MTIME = 0L;
```

File: clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java [hunk 2]
```diff
@@ -289,41 +288,191 @@ public void testAppend() throws IOException {
         fs.append(null, 0, null);
     }
 
-    /*
+    private void mockNonExistingPath(ObjectLocation objectLoc) throws ApiException {
+        when(objectsApi.statObject(objectLoc.getRepository(), objectLoc.getRef(), objectLoc.getPath()))
+                .thenThrow(new ApiException(HttpStatus.SC_NOT_FOUND, "no such file"));
+        when(objectsApi.listObjects(objectLoc.getRepository(), objectLoc.getRef(),
+                objectLoc.getPath() + Constants.SEPARATOR, "", 1, ""))
```
Comment (arielshaqed) [inline clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java:295]: AFAICT the "1" here should be an `any()` (and then, unfortunately, Mockito will probably want all the parameters to be `eq("")` and similar): *any* file listing should return nothing, no matter how many entries I ask for.

Comment (arielshaqed) [inline clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java:295]: Also I don't understand the trailing `Constants.SEPARATOR`: if you specify that `a/b/c` does not exist, I would expect that listing `a/b/c` would _also_ not exist.

File: clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java [hunk 3]
```diff
@@ -289,41 +288,191 @@ public void testAppend() throws IOException {
...
+                .thenReturn(new ObjectStatsList());
+    }
+
+    private void mockExistingDirPath(ObjectLocation dirObjLoc, ObjectLocation fileInDir) throws ApiException {
```
Comment (arielshaqed) [inline clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java:299]: Please document this function, it is unclear to a reader like myself what the parameters mean & what it does.

File: clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java [hunk 4]
```diff
@@ -289,41 +288,191 @@ public void testAppend() throws IOException {
...
+        return srcStats;
+    }
+
+    private String objectLocToS3ObjKey(ObjectLocation objectLoc) {
+        return String.format("/%s/%s/%s",objectLoc.getRepository(), objectLoc.getRef(), objectLoc.getPath());
+    }
+
+
+    private void verifyObjDeletion(ObjectLocation srcObjLoc) throws ApiException {
+        verify(objectsApi).deleteObject(eq(srcObjLoc.getRepository()), eq(srcObjLoc.getRef()), eq(srcObjLoc.getPath()));
```
Comment (arielshaqed) [inline clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java:334]: Why do we need matchers if they are all `eq()`?
```suggestion
        verify(objectsApi).deleteObject(srcObjLoc.getRepository(), srcObjLoc.getRef(), srcObjLoc.getPath());
```

Comment (arielshaqed) [reply clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java:334]: I don't think that this function abstracts anything and actually _prefer_ the explicit version.

File: clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java [hunk 5]
```diff
@@ -289,41 +288,191 @@ public void testAppend() throws IOException {
...
+        verifyObjDeletion(fileObjLoc);
+    }
+
+    /**
+     * rename(srcDir(containing srcDir/a.txt), existing-dstdir) -> existing-dstdir/srcDir/a.txt
+     */
+    @Test
+    public void testRename_existingDirToExistingDirName() throws ApiException, IOException {
+        Path fileInSrcDir = new Path("lakefs://repo/main/existing-dir1/existing.src");
+        ObjectLocation srcFileObjLoc = fs.pathToObjectLocation(fileInSrcDir);
```
Comment (arielshaqed) [inline clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java:434]: 1. Can we have multiple files in the directory here?
2. Should/can we have more than one page's worth of files here (so we can see pagination at work)?

**Review by arielshaqed — CHANGES_REQUESTED** [review]
Cool!

Nothing major, mostly requesting comments and a bit more coverage of multi-file renames.

File: clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java (hunk 2 above)
Comment (talSofer) [reply clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java:295]: The problem is that Mockito does not allow mixing matcher (e.g. any()) with explicit parameters..

File: clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java [hunk 6]
```diff
@@ -54,15 +54,14 @@
...
-import io.lakefs.clients.api.model.ObjectStats;
 import io.lakefs.clients.api.model.ObjectStats.PathTypeEnum;
-import io.lakefs.clients.api.model.ObjectStatsList;
-import io.lakefs.clients.api.model.Repository;
-import io.lakefs.clients.api.model.StagingLocation;
-import io.lakefs.clients.api.model.StagingMetadata;
 
 public class LakeFSFileSystemTest {
     protected static final Logger LOG = LoggerFactory.getLogger(LakeFSFileSystemTest.class);
+    private static final Long FILE_SIZE = 1L;
```
Comment (arielshaqed) [inline clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java:61]:
```suggestion
    private static final Long UNUSED_FILE_SIZE = 1L;
```

File: clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java [hunk 7]
```diff
@@ -289,41 +288,215 @@ public void testAppend() throws IOException {
...
+
+        Path dst = new Path("lakefs://repo/another-branch/existing.dst");
+        ObjectLocation dstObjLoc = fs.pathToObjectLocation(dst);
+
+        boolean renamed = fs.rename(src, dst);
+        Assert.assertFalse(renamed);
+    }
+
+    /**
+     * no-op. rename is expected to succeed.
```
Comment (arielshaqed) [inline clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java:475]: Can we verify that nothing happened?  E.g. `verify()` on staging that neither operation (get URL for upload, link URL) was tried.

File: clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java [hunk 8]
```diff
@@ -289,41 +288,215 @@ public void testAppend() throws IOException {
...
+    @Test
+    public void testRename_srcAndDstOnDifferentBranch() throws IOException {
+        Path src = new Path("lakefs://repo/branch/existing.src");
+        ObjectLocation srcObjLoc = fs.pathToObjectLocation(src);
+
+        Path dst = new Path("lakefs://repo/another-branch/existing.dst");
+        ObjectLocation dstObjLoc = fs.pathToObjectLocation(dst);
+
+        boolean renamed = fs.rename(src, dst);
+        Assert.assertFalse(renamed);
```
Comment (arielshaqed) [inline clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java:471]: Here we might also want to verify that no API operations were performed on staging.

File: clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java [hunk 9]
```diff
@@ -291,33 +291,38 @@ public void testAppend() throws IOException {
...
-        when(objectsApi.listObjects(objectLoc.getRepository(), objectLoc.getRef(),
-                objectLoc.getPath() + Constants.SEPARATOR, "", 1, ""))
+
+        when(objectsApi.listObjects(eq(objectLoc.getRepository()), eq(objectLoc.getRef()),
+                eq(objectLoc.getPath() + Constants.SEPARATOR), eq(""), any(), eq("")))
                 .thenReturn(new ObjectStatsList());
     }
 
-    private void mockExistingDirPath(ObjectLocation dirObjLoc, ObjectLocation fileInDir) throws ApiException {
+    private void mockExistingDirPath(ObjectLocation dirObjLoc, ArrayList<ObjectLocation> filesInDir) throws ApiException {
```
Comment (arielshaqed) [inline clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java:300]:
```suggestion
    private void mockExistingDirPath(ObjectLocation dirObjLoc, List<ObjectLocation> filesInDir) throws ApiException {
```

File: clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java [hunk 10]
```diff
@@ -394,7 +398,7 @@ public void testRename_existingFileToExistingDirName() throws ApiException, IOEx
         ObjectLocation fileObjLoc = fs.pathToObjectLocation(fileInDstDir);
         Path dst = new Path("lakefs://repo/main/existing-dir2");
         ObjectLocation dstObjLoc = fs.pathToObjectLocation(dst);
-        mockExistingDirPath(dstObjLoc, fileObjLoc);
+        mockExistingDirPath(dstObjLoc, new ArrayList<ObjectLocation>(Arrays.asList(fileObjLoc)));
```
Comment (arielshaqed) [inline clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java:401]:
```suggestion
        mockExistingDirPath(dstObjLoc, ImmutableList.of(fileObjLoc));
```
(Also below, I think we can get by without `Arrays` and directly generate a list).

Alternatively, pass an array and iterate over _that_.  Mixing Lists and native arrays is too verbose if we don't have to do it.

**Review by arielshaqed — APPROVED** [review]
Thanks, great tests!  But please do go over comments, in particular I think we should pick arrays or Lists and stay with that on both sides of a test-internal API.

**Pull request merged**